sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.ordered.combiner;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.extensions.ordered.ContiguousSequenceRange;
+import org.apache.commons.lang3.tuple.Pair;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.UnknownKeyFor;
+import org.joda.time.Instant;
+
+/** Default accumulator used to combine sequence ranges. */
+public class SequenceRangeAccumulator {
+
+  private static Instant max(Instant a, Instant b) {
+    return a.isAfter(b) ? a : b;
+  }
+
+  private final TreeMap<Long, Pair<Long, Instant>> data = new TreeMap<>();
+  private @Nullable Long initialSequence = null;
+
+  public void add(long sequence, Instant timestamp, boolean isInitialSequence) {
+    if (isInitialSequence && this.initialSequence != null && sequence != this.initialSequence) {
+      throw new IllegalStateException(
+          "There are different initial sequences detected: "
+              + initialSequence
+              + " and "
+              + sequence);
+    }
+
+    if (isInitialSequence) {
+      this.initialSequence = sequence;
+      clearRangesBelowInitialSequence(sequence, timestamp);
```

Comment:
When would we have ranges which are below the initial sequence? Do we need this, or should we just throw? I'd expect this to be an error case...

Reply:
It can be the case of data reprocessing when customers need to re-play a number of messages but the source doesn't support the exact rewinding (e.g. Pub/Sub only allows time based reset to a snapshot). Also, there could be use cases where the initial sequence is not known upfront and is derived from some message data. In that case we are only interested in the range from the starting sequence.

Reply:
Ack - maybe we should at least warn? In this case we are intentionally discarding data, which might be unexpected

Reply:
The data is not discarded in this case; it's output to the DLQ output with "duplicate" reason - ProcessorDoFn, line 349 (for some reason can't create a permalink in PR).

Reply:
Oh I see, you're right. Thanks